File: src/problems1_9/Problem4.java
package problems1_9;

/**
 * @author chris
 *         <p/>
 *         <p/>
 *         A palindromic number reads the same both ways. The largest palindrome
 *         made from the product of two 2-digit numbers is 9009 = 91 × 99.
 *         <p/>
 *         Find the largest palindrome made from the product of two 3-digit
 *         numbers.
 */
public class Problem4 {
    public static void main(String[] args) {
        int highest = 0;
        final long t1 = System.currentTimeMillis();
        for (int a = 100; a < 1000; a++) {
            for (int b = a + 1; b < 1000; b++) {
                if (isPalindrome(Integer.toString(a * b))) {
                    if (a * b > highest) {
                        highest = a * b;
                    }
                }
            }
        }
        final long t2 = System.currentTimeMillis();
        System.out.println(highest);
        System.out.println("Execution time: " + (double) (t2 - t1) / 1000);
    }

    static boolean isPalindrome(String cheese) {
        String reverse = "";
        for (int i = 0; i < cheese.length(); i++) {
            reverse += cheese.substring(cheese.length() - 1 - i,
                    cheese.length() - i);
        }
        return reverse.equals(cheese);
    }
}

File: problems01_09/Problem4.cc
#include <stdio.h>

int isPalindrome(unsigned n){
  char num_storage[16]; // should be < 16 digit number
  unsigned count = 0;
  while(n){
    num_storage[count++] = n%10;
    n/=10;
  }
  for(unsigned i = 0; i < count/2;++i)
    if (num_storage[i] != num_storage[count-i-1])
      return 0;
  return 1;
}

int main(){
  int highest = 0;
  for(int i = 100; i < 1000; ++i){
    for(int j = i; j < 1000; ++j){
      if (i*j > highest && isPalindrome(i*j))
	highest = i*j;
    }
  }
  printf("%d\n", highest);
}
